Add behaviour for no tags present

shestee opened the pull request:
In AutoVersion in catch block the presence of the Optional was not checked; now it deducts version based on requestedVersion from config, and if it is not present `orElse()` returns fallback value for version: `0.0.1-SNAPSHOT`. This fixes #82 and also resolves #83 but additionally tag info finding is moved to `try-catch` block in `pickNextVersion()` - when an exception is caught fallback value for version is returned and more detailed explanation (than general one in AutoVersion) is logged. It is now also more visible in `NextVersionPicker` that `run("git", "describe", "--tags")` can throw an Exception.

shestee committed:
Added behaviour for no tags present

When 'git describe --tags' can't find a tag, fallback version ('0.0.1-SNAPSHOT') is deducted.

koral-- commented on src/main/java/org/shipkit/auto/version/NextVersionPicker.java:
```diff
@@ -33,8 +33,17 @@ String pickNextVersion(Optional<Version> previousVersion, VersionConfig config,
+            try {
+                tag = runner.run("git", "describe", "--tags").trim();
+            } catch (Exception e) {
+                result = "0.0.1-SNAPSHOT";
```
I think that exception should be somehow utilized. By logging it perhaps.

mockitoguy replied:
Good point. I think we should add both:
 - `info` level clean message
 - `debug` level message with stack trace

```
log.info("Process 'git describe --tags' exited with error. Assuming there are no tags. Run with --debug for more.");
log.debug("Ignored exception from 'git describe --tags'. Assuming there are no tags.", e);
```

@shestee, do you want to add those?

shestee replied:
> I think that exception should be somehow utilized. By logging it perhaps.

Thanks for suggestion @koral-- ! You're right. I think logging the exception is fine here. I add `log.info()` and `log.debug()` as @mockitoguy proposed.

mockitoguy commented on src/main/java/org/shipkit/auto/version/NextVersionPicker.java:
```diff
@@ -33,8 +33,17 @@ String pickNextVersion(Optional<Version> previousVersion, VersionConfig config,
+                tag = runner.run("git", "describe", "--tags").trim();
+            } catch (Exception e) {
+                result = "0.0.1-SNAPSHOT";
+                explainVersion(log, result, "couldn't find tags in project");
```
"couldn't find tags in project" -> "couldn't run 'git describe --tags' (were the tags fetched?)"

mockitoguy reviewed (changes requested):
I suggest a few changes and we should be ready to merge. Thanks!

shestee committed:
Added an exception logging in NextVersionPicker

'run("git", "describe", "--tags")' can throw exception which is logged when caught.

mockitoguy commented on src/main/java/org/shipkit/auto/version/NextVersionPicker.java:
```diff
@@ -33,8 +33,20 @@ String pickNextVersion(Optional<Version> previousVersion, VersionConfig config,
+                log.info("Process 'git describe --tags' exited with non-zero exit value. Assuming there are no tags. " +
+                        "Run with --debug for more.");
+                log.debug("Ignored exception from 'git describe --tags'. Assuming there are no tags.", e);
+                explainVersion(log, result, "couldn't run 'git describe --tags'");
```
"couldn't run 'git describe --tags'" -> "couldn't run 'git describe --tags' (assuming there are no tags)"

shestee replied:
@mockitoguy thanks, it's done

mockitoguy reviewed (changes requested):
One little change and we're ready to merge!
